File: src/string-and-array/container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int ans = 0;

        int l = 0;
        int r = height.size() - 1;

        while(l < r) {
            int minH = min(height[l], height[r]);
            ans = max(ans, minH * (r - l));

            while (height[l] <= minH && l < r) {
                l++;
            }

            while (height[r] <= minH && l < r) {
                r--;
            }
        }

        return ans;
    }
};

File: src/string-and-array/container-with-most-water.rs
impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let mut l = 0usize;
        let mut r = height.len() - 1;
        let mut ans: i32 = 0;
        while (l < r) {
            ans = i32::max(ans, (r - l) as i32 * i32::min(height[l], height[r]));
            if height[l] < height[r] {
                l = l + 1;
            } else {
                r = r - 1;
            }
        }

        return ans;
    }
}